### PR TITLE
sql: sequences created for identity columns have incorrect types

### LINKED ISSUE #110352: sql: incorrect sequence type and max value created for Identity column
**Describe the problem**

Given a table created using the following SQL

```
CREATE TABLE public."People" (
  "Id" INT4 NOT NULL GENERATED BY DEFAULT AS IDENTITY,
  "SomeColumn" INT4 NOT NULL GENERATED BY DEFAULT AS IDENTITY,
  CONSTRAINT "PK_People" PRIMARY KEY ("Id" ASC)
)
```

```
SELECT
  sequence_schema, sequence_name,
  data_type AS seqtype,
  start_value::bigint AS seqstart,
  minimum_value::bigint AS seqmin,
  maximum_value::bigint AS seqmax,
  increment::bigint AS seqincrement,
  1::bigint AS seqcache,
  CASE
    WHEN cycle_option = 'YES' THEN TRUE
    ELSE FALSE
  END AS seqcycle
FROM information_schema.sequences
JOIN pg_namespace AS ns ON ns.nspname = sequence_schema
JOIN pg_class AS cls ON cls.relnamespace = ns.oid AND cls.relname = sequence_name;

```

the sequence that backs the Id column is as follows

```
  sequence_schema |     sequence_name     | seqtype | seqstart | seqmin |   seqmax   | seqincrement | seqcache | seqcycle
------------------+-----------------------+---------+----------+--------+------------+--------------+----------+-----------
  public          | People_SomeColumn_seq | INT8    |        1 |      1 | 2147483647 |            1 |        1 |    f
  public          | People_Id_seq         | INT8    |        1 |      1 | 2147483647 |            1 |        1 |    f
```

Issues:

- The sequence type INT8 while Id column is INT4
- The seqmax only 2147483647 (= max int 32)

**Additional context**
What was the impact?

This affects efcore.pg tests

@fqazi 

Jira issue: CRDB-31394

### PR DESCRIPTION
**Pull request by fqazi**
Previously, when a sequence was created for an identity column, the type was not correctly propagated. This would cause those sequences to always be as of type Int. This behaviour was incorrect; they should match the type based on the column. To address, this patch will propagate the type into the sequence option.

Fixes: #110352

Release note (bug fix): CREATE TABLE with IDENTITY columns did not properly propagate the type of the column into the sequence.

### CODE REVIEW

File: pkg/sql/serial.go [hunk 1]
```diff
@@ -141,6 +141,7 @@ func (p *planner) generateSerialInColumnDef(
 	if err != nil {
 		return nil, nil, nil, nil, err
 	}
+	seqTyp := defType
```
Comment (ecwall) [inline pkg/sql/serial.go:144]: It is confusing that there are 2 similarly named variables `seqTyp` and `seqType`. Could you rename one or both to make it clear what a "type" is in the context of a sequence?

Comment (fqazi) [reply pkg/sql/serial.go]: Good point, and I renamed it to asIntType, since it maps to the AS <TYPENAME> clause on the create sequence.

**Review by ecwall — APPROVED** [review]

**Comment by ecwall** [comment]
@fqazi, does this need to be backported?

File: pkg/sql/serial.go [hunk 2]
```diff
@@ -211,6 +213,23 @@ func (p *planner) generateSerialInColumnDef(
...
 		}
 	}
+
+	// Setup the type of the sequence based on the type observed within
+	// the column.
+	switch asIntType {
+	case types.Int2, types.Int4:
+		// Valid types, nothing to do.
+	case types.Int:
+		// Int is the default, so no cast necessary.
```
Comment (rafiss) [inline pkg/sql/serial.go:223]: does this correctly handle the case where `default_int_size` is changed?

Comment (fqazi) [reply pkg/sql/serial.go:223]: Ah, have to handle that case. Let me adjust the code

**Review by fqazi — COMMENTED** [review]
Yes, we need that for efcore. Let me tag it

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110621#-:-NeoSxlk7tu4MIjfH3DV:bsnedcw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[`pkg/sql/serial.go` line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110621#gh-1331896865:-NeoSvRb0lQOyInTK2jS:buqpn7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/087ddf793cf045f3f103e6ed682781536434ecdb/pkg/sql/serial.go#L223)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Ah, have to handle that case. Let me adjust the code
</blockquote></details>

Added test coverage only, this works correctly.

File: pkg/sql/logictest/testdata/logic_test/create_table [hunk 3]
```diff
@@ -983,3 +983,59 @@ tbl  CREATE TABLE public.tbl (
        CONSTRAINT tbl_pkey PRIMARY KEY (i ASC),
        FAMILY f1 (i, j)
      )
+
+# We should propogate the type of the column into the sequence (#110352), instead
+# of always using INT8 for them.
+subtest regression_43894
```
Comment (rafiss) [inline pkg/sql/logictest/testdata/logic_test/create_table:989]: super nit: the `regression_43894` name should be updated, i think

**Review by rafiss — APPROVED** [review]

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110621)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[`pkg/sql/logictest/testdata/logic_test/create_table` line 989 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110621#gh-1332221742:-NepAyzS7rSQD9EcKn4V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eaf23d6214d70703516525d0db3f48af2bc68b45/pkg/sql/logictest/testdata/logic_test/create_table#L989)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: the `regression_43894` name should be updated, i think
</blockquote></details>

Done.

**Pull request merged**